fdbserver/RocksDBCheckpointUtils.actor.cpp

```diff
@@ -20,32 +20,398 @@
+
+			double getTimeEstimate() const override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
+
+			const std::string path;
+			const KeyRange range;
+			const Version version;
+			ThreadReturnPromise<Void> done;
+		};
+
+		struct CloseAction : TypedAction<Reader, CloseAction> {
+			CloseAction(std::string path, bool deleteOnClose) : path(path), deleteOnClose(deleteOnClose) {}
+			double getTimeEstimate() const override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
+
+			std::string path;
+			bool deleteOnClose;
+			ThreadReturnPromise<Void> done;
+		};
+
+		struct ReadRangeAction : TypedAction<Reader, ReadRangeAction>, FastAllocated<ReadRangeAction> {
+			ReadRangeAction(int rowLimit, int byteLimit)
+			  : rowLimit(rowLimit), byteLimit(byteLimit), startTime(timer_monotonic()) {}
+
+			double getTimeEstimate() const override { return SERVER_KNOBS->READ_RANGE_TIME_ESTIMATE; }
+
+			const int rowLimit, byteLimit;
+			const double startTime;
+			ThreadReturnPromise<RangeResult> result;
+		};
+
+		explicit Reader(DB& db);
+		~Reader() override {}
+
+		void init() override {}
+
+		void action(OpenAction& a);
+
+		void action(CloseAction& a);
+
+		void action(ReadRangeAction& a);
+
+		DB& db;
+		CF cf;
+		Key begin;
+		Key end;
+		double readRangeTimeout;
+		std::unique_ptr<rocksdb::Iterator> cursor;
+	};
+
+	ACTOR static Future<Void> doClose(RocksDBCheckpointReader* self);
+
+	DB db = nullptr;
+	std::string path;
+	const UID id;
+	Version version;
+	Reference<IThreadPool> readThreads;
+	Future<Void> openFuture;
+};
+
+RocksDBCheckpointReader::RocksDBCheckpointReader(const CheckpointMetaData& checkpoint, UID logID)
+  : id(logID), version(checkpoint.version) {
+	RocksDBCheckpoint rocksCheckpoint = getRocksCheckpoint(checkpoint);
+	this->path = rocksCheckpoint.checkpointDir;
+	if (g_network->isSimulated()) {
+		readThreads = CoroThreadPool::createThreadPool();
+	} else {
+		readThreads = createGenericThreadPool();
+	}
+	readThreads->addThread(new Reader(db), "fdb-rocks-rd");
+}
+
+Future<Void> RocksDBCheckpointReader::init(StringRef token) {
+	if (openFuture.isValid()) {
+		return openFuture;
+	}
+
+	KeyRange range = BinaryReader::fromStringRef<KeyRange>(token, IncludeVersion());
+	auto a = std::make_unique<Reader::OpenAction>(this->path, range, this->version);
+	openFuture = a->done.getFuture();
+	readThreads->post(a.release());
+	return openFuture;
+}
+
+Future<RangeResult> RocksDBCheckpointReader::nextKeyValues(const int rowLimit, const int byteLimit) {
+	auto a = std::make_unique<Reader::ReadRangeAction>(rowLimit, byteLimit);
+	auto res = a->result.getFuture();
+	readThreads->post(a.release());
+	return res;
+}
+
+RocksDBCheckpointReader::Reader::Reader(DB& db) : db(db), cf(nullptr) {
+	if (g_network->isSimulated()) {
+		// In simulation, increasing the read operation timeouts to 5 minutes, as some of the tests have
+		// very high load and single read thread cannot process all the load within the timeouts.
```

Comment:
I agree the slowness worth more investigation. I copied this from the `KeyValueStoreRocksDB.actor.cpp`. There is currently `FlowLock`s used in `KeyValueStoreRocksDB` to limit the number of reads in the queue. And in this CheckpointReader, this will be a single reader thread. The slowness will be investigated, though likely not in this PR.